src/server/req_delete.c

```diff
@@ -236,68 +236,107 @@ check_deletehistoryjob(struct batch_request * preq)
-	if (historyjob == PBSE_HISTJOBID) {
+	if (svr_chk_histjob(pjob) != PBSE_HISTJOBID) {
+		/*
+		 *  Job is in a Non Finished state . It must be terminated and then its history
+		 * should be purged .
+		 */
+		return FALSE;
+	}
+
+	/* Issue history job delete request to remote server if job is moved. */
+	if (pjob->ji_qs.ji_state == JOB_STATE_MOVED)
+		issue_delete(pjob, preq);
+
+	if (jt == IS_ARRAY_NO) {
 		snprintf(log_buffer, sizeof(log_buffer),
 			msg_job_history_delete, preq->rq_user,
 			preq->rq_host);
 		log_event(PBSEVENT_DEBUG, PBS_EVENTCLASS_JOB, LOG_INFO,
 			jid,
 			log_buffer);
 
-		/* Issue history job delete request to remote server if job is moved. */
-		if (histpjob->ji_qs.ji_state == JOB_STATE_MOVED)
-			issue_delete(histpjob);
+		job_purge(pjob);
+		preq->rq_reply.brp_code = PBSE_HISTJOBDELETED;
+		reply_send(preq);
+		return TRUE;
+	} else if (jt == IS_ARRAY_ArrayJob) {
+
+		/* The Array Job itself ... */
+		/* for each subjob of array do purge */
+
+		if (pjob->ji_ajtrk) {
+			for (i = 0; i < pjob->ji_ajtrk->tkm_ct; i++) {
+				sjid = mk_subjob_id(pjob, i);
+
+				psjob = find_job(sjid);
 
-		job_purge(histpjob);
+				if (psjob) {
+					snprintf(log_buffer, sizeof(log_buffer),
+						msg_job_history_delete, preq->rq_user,
+						preq->rq_host);
+					log_event(PBSEVENT_DEBUG, PBS_EVENTCLASS_JOB, LOG_INFO,
+						sjid,
+						log_buffer);
+
+					job_purge(psjob);
+				}
+			}
+		}
+
+		snprintf(log_buffer, sizeof(log_buffer),
+			msg_job_history_delete, preq->rq_user,
+			preq->rq_host);
+		log_event(PBSEVENT_DEBUG, PBS_EVENTCLASS_JOB, LOG_INFO,
+			jid,
+			log_buffer);
+
+		job_purge(pjob);
 		preq->rq_reply.brp_code = PBSE_HISTJOBDELETED;
 		reply_send(preq);
 		return TRUE;
-	} else {
-		/*
-		 *  Job is in a Non Finished state . It must be terminated and then its history
-		 * should be purged .
-		 */
-		return FALSE;
 	}
+
+	return FALSE;
 }
 
 /**
  * @Brief
  *		Issue PBS_BATCH_DeleteJob request to remote server.
  *
- * @param[in]	pjob - Job structure.
+ * @param[in]	preq	- Batch request structure.
  */
 void
-issue_delete(pjob)
-job *pjob;
+issue_delete(job *pjob, struct batch_request   *preq)
 {
-	struct batch_request   *preq;
+	struct batch_request  *npreq;
 	char rmt_server[PBS_MAXSERVERNAME + 1] = {'\0'};
 	char *at = NULL;
 
 	if (pjob == NULL)
 		return;
+	if (preq == NULL)
+		return;
 
 	if ((at = strchr(pjob->ji_wattr[JOB_ATR_in_queue].at_val.at_str, (int)'@')) == NULL)
 		return;
 
 	strncpy(rmt_server, at + 1, PBS_MAXSERVERNAME + 1);
 
-	preq = alloc_br(PBS_BATCH_DeleteJob);
-	if (preq == NULL)
+	npreq = alloc_br(PBS_BATCH_DeleteJob);
+	if (npreq == NULL)
 		return;
 
-	(void)strcpy(preq->rq_ind.rq_delete.rq_objname, pjob->ji_qs.ji_jobid);
-	preq->rq_extend = (char *)malloc(strlen(DELETEHISTORY) + 1);
-	if (preq->rq_extend == NULL) {
+	(void)strcpy(npreq->rq_ind.rq_delete.rq_objname, preq->rq_ind.rq_delete.rq_objname);
+	npreq->rq_extend = (char *)malloc(strlen(DELETEHISTORY) + 1);
```

Comment:
No need to typecast malloc here.

Reply:
Removed the typecast in rc 6 https://github.com/PBSPro/pbspro/pull/590/commits/ad3e2f2db9b9fb0b0bf8e4be9030b950412dc17e   All though the typecast was not introduced with this PR